tools/codegen/README.md

```diff
@@ -78,3 +74,100 @@ update-codegen-crds:
 
 You may also want to add the `_output` directory to your `.gitignore` to avoid checking in compiled binaries created
 by this make target.
+
+## Generators
+
+The following section describes the individual generators included within the `codegen` utility.
+
+The generators enabled by default are:
+- [Compatibility](#compatibility)
+- [Schemapatch](#schemapatch)
+- [Swaggerdocs](#swaggerdocs)
+
+### Compatibility
+
+To generate API compatibility comments, use the `compatibility` subcommand.
+
+The `compatibility` subcommand generates a compatibility level comment for each API defintion.
+The generation is controlled by a marker applied to the CRD struct defintiion.
+For example, this annotation would be `+openshift:compatibility-gen:level=1` for a level 1 API.
+	
+Valid API levels are 1, 2, 3 and 4. Version 1 is required for all stable APIs.
+Version 2 should be used for beta level APIs. Levels 3 and 4 may be used for alpha APIs.
+
+### Schemapatch
+
+To generate CRD schemas, use the `schemapatch` subcommand.
+
+The schemapatch subcommand uses the [controller-tools][controller-tools] project to generate CRD schemas.
+It requires a stub CRD to exist before the schema will be generated. Copying an existing OpenShift CRD
+and updating the metadata is the easiest way to generate a stub.
+
+In addition to generating schemas, it also allows patches to be applied to the schema post generation.
+The [yaml-patch][yaml-patch] library accepts RFC-6092-ish patches in a YAML format and applies them to
+the CRD.
+YAML patch files must be named identically to the CRD they apply to, but with a `yaml-patch` extension
+instead of the standard `yaml` extension.
+
+These additional arguments may be provided when using the `schemapatch` generator:
+- `--controller-gen` - optionally use a particular controller-gen binary. When not specified, the tool will use the
+  built in generator.
+  Note, you must use a `controller-gen` built from the [OpenShift fork](https://github.com/openshift/kubernetes-sigs-controller-tools) of controller-tools.
+- `--required-feature-sets` - optionally generate based on the OpenShift feature sets annotations.
```

Comment:
do we have an eventual goal to stop allowing this to be set?  Why only generate some of the schemas?

Reply:
I introduced this originally because of how the feature sets are being handled in the controller-gen tool, where it can only check against a single value for `OPENSHIFT_REQUIRED_FEATURESETS`.

I can see a future iteration where, when not specified, we work out which feature set combinations exist within the API group and then generate them all